Show Name of Undefined Control Sequence (macro)

Strauman opened the pull request:
When LaTeX complains about undefined control sequence, this plugin does
not show the name of said macro, which is what this commit attempts to
fix.

Strauman committed:
Show Name of Undefined Control Sequence (macro)

When LaTeX complains about undefined control sequence, this plugin does
not show the name of said macro, which is what this commit attempts to
fix.

Strauman committed:
Show Name of Undefined Control Sequence (macro)

When LaTeX complains about undefined control sequence, this plugin does
not show the name of said macro, which is what this commit attempts to
fix.

Strauman commented:
Simple LaTeX test case:

```
\documentclass{minimal}
\def\once{Once}
\begin{document}
  \once upon a time there was an \undefined control sequence.
\end{document}
```

Expected log output:
<img width="366" alt="screen shot 2018-11-11 at 11 52 27" src="https://user-images.githubusercontent.com/1666236/48311987-4afa8a80-e5a8-11e8-8b71-596976d2f257.png">

Strauman commented on lib/parsers/log-parser.js:
```diff
@@ -18,6 +18,12 @@ const ERROR_PATTERN = new RegExp('' +
+// Pattern for obtaining the last macro in a line
+// Meant for catching the undefined control sequence
+const LAST_MACRO_PATTERN = new RegExp('' +
+  '[^\\\\]*(\\\\[a-zA-Z@]+)$' // Macro name
```
I'm not sure as to what would constitute a legal macro name. 

One could argue that it _might_ be better to just look at the last
space and go from there.

An issue with the definition here is that I din't include underscores `_`, 
which will be very widely used in macro names come LaTeX3.

```suggestion
  '[^\\\\]*(\\\\[a-zA-Z@_]+)$' // Macro name
```

Strauman committed:
Update log-parser.js

Strauman committed:
Update log-parser.js

Strauman commented on lib/parsers/log-parser.js:
```diff
@@ -18,6 +18,12 @@ const ERROR_PATTERN = new RegExp('' +
+// Pattern for obtaining the last macro in a line
+// Meant for catching the undefined control sequence
+const LAST_MACRO_PATTERN = new RegExp('' +
+  '(\\\\[a-zA-Z@]+)$' // Macro name
```
I'm not confident as to what ideally should constitute a legal macro name in
this setting.

One of the biggest gotchas here is that in LaTeX, even space can be made
a letter-character (thus allowed in macro names). However, that would be
weirdly inconvenient. The other is that any character can be made catcode 0,
that is, the escape character (by default `\`).

I would argue that it's be better to just look at the last space and
go from there, as it is rather unlikely that one would redefine the space
as such:

```suggestion
  '([^ ]+)$' // Macro name
```

An issue with the definition here is that I din't include underscores `_`, 
which will be very widely used in macro names come LaTeX3.

```suggestion
  '(\\\\[a-zA-Z@_]+)$' // Macro name
```

Strauman replied:
Actually, I think the best way would be to assume that `\` is the escapechar and

```suggestion
  '(\\\\[^\\\\]+)$' // Macro name
```

Strauman committed:
Increase Timeout on Kill Latexmk Spec

Strauman committed:
Increase Timeout on Kill Latexmk Spec

Strauman committed:
Merge branch 'control-sequence-name' of github.com:Strauman/atom-latex into control-sequence-name

Strauman commented:
Probably a test should be made to see if there is any matches!

Strauman committed:
Switch to Better RegExp for LAST_MACRO_PATTERN

Strauman committed:
Switch to Better RegExp for LAST_MACRO_PATTERN

Strauman committed:
Change Behaviour If No Match in Undefined Macro

Also changed `var`s to `let` in log-parser.js

Strauman committed:
Change Behaviour If No Match in Undefined Macro

Also changed `var`s to `let` in log-parser.js

Strauman commented on lib/parsers/log-parser.js:
```diff
@@ -18,6 +18,12 @@ const ERROR_PATTERN = new RegExp('' +
+// Pattern for obtaining the last macro in a line
+// Meant for catching the undefined control sequence
+const LAST_MACRO_PATTERN = new RegExp('' +
+  '(\\\\[^\\\\]+)$' // Macro name
```
I'm not highly confident as to what ideally should a legal macro name in
this setting.

The problem is that the escape character, `\` by default, can be redefined, as well
as having anything being active letters (and thus treated as control sequences)

E.g. doing
```
\catcode`\O=13\relax O
```
Would make the letter O an unknown control sequence, or

```
\catcode`J=0\relaxJUndefined
```
Would make `JUnedfined` behave identically to `\Undefined` and this atom plugin
will report weird information.

You can't just look at the last space and go from there. If you consider
```
{}\some@undef@macro
```
Then the `{}` would be a part of the macro name.

However, I do argue that these cases are extraordinary circumstances,
and isn't really that necessary to implement.

Strauman replied:
I'm not highly confident as to what ideally should a legal macro name in
this setting.

The problem is that the escape character, `\` by default, can be redefined, as well
as having anything being active letters (and thus treated as control sequences)

E.g. doing
```
\catcode`O=13\relax O
```
would make the letter O an unknown control sequence, and

```
\catcode`J=0\relax JUndefined
```
would make `JUnedfined` behave identically to `\Undefined` and this atom plugin
will report weird information.

You can't just look at the last space and go from there. If you consider
```
{}\some@undef@macro
```
Then the `{}` would be a part of the macro name.

However, I do argue that these cases are extraordinary circumstances,
and isn't really that necessary to implement.

Strauman commented on lib/parsers/log-parser.js:
```diff
@@ -76,10 +85,27 @@ export default class LogParser extends Parser {
+        let outMessageText
         const lineNumber = match[2] ? parseInt(match[2], 10) : undefined
+        // Check if the error is "Undefined control sequence"
+        var errMessageText = match[4]
```
Sorry 😰

```suggestion
        let errMessageText = match[4]
```

Strauman committed:
Use Let Instead Of Var

Strauman committed:
Use Let Instead Of Var

thomasjo commented on lib/parsers/log-parser.js:
```diff
@@ -76,10 +85,27 @@ export default class LogParser extends Parser {
+          // The actual macro is the last thing
+          // in the sentence of the next line
+          let nextLine = lines[index + 1]
+          let macroMatch = nextLine.match(LAST_MACRO_PATTERN)
```
Let's declare `errMessageText`, `nextLine`, and `macroMatch` as `const`.

thomasjo commented on lib/parsers/log-parser.js:
```diff
@@ -76,10 +85,27 @@ export default class LogParser extends Parser {
 
       match = line.match(ERROR_PATTERN)
       if (match) {
+        let outMessageText
```
I think maybe this variable should be named `outputMessage`. The `Text` suffix is a bit redundant.

Strauman replied:
Agreed!

thomasjo commented on lib/parsers/log-parser.js:
```diff
@@ -76,10 +85,27 @@ export default class LogParser extends Parser {
+        let outMessageText
         const lineNumber = match[2] ? parseInt(match[2], 10) : undefined
+        // Check if the error is "Undefined control sequence"
+        let errMessageText = match[4]
```
What do you think about renaming this to `errorMessage`?

Strauman replied:
Yes. I am a fan of explicit names. `errorMessage` it is!

thomasjo commented on lib/parsers/log-parser.js:
```diff
@@ -76,10 +85,27 @@ export default class LogParser extends Parser {
+        if (errMessageText === UNDEF_CONTROL_SEQ_TEXT) {
+          // The actual macro is the last thing
+          // in the sentence of the next line
+          let nextLine = lines[index + 1]
```
I guess in most situations this will be safe given how the error is logged, but if we wanted to be super safe we should probably check that this index lookup is not out of bounds. One possibility is to extend the condition above like

``` js
if (errMessageText === UNDEF_CONTROL_SEQ_TEXT && lines.length > index) {
```

Although that might be wrong place to short-circuit this depending upon what error message you want to give to the user in that case. If you're pretty sure we'll never go out of bounds here, we can skip this check.

Strauman replied:
I can't think of a scenario where the the error would be `Undefined control sequence` and no control sequence will follow in the line below. However, I'll do an out-of-bounds check, to be sure?

thomasjo commented on lib/parsers/log-parser.js:
```diff
@@ -76,10 +85,27 @@ export default class LogParser extends Parser {
+          let macroMatch = nextLine.match(LAST_MACRO_PATTERN)
+          // Check if there is a match
+          if (macroMatch) {
+            outMessageText = errMessageText + ': ' + macroMatch[1]
```
Replace this with

``` js
outMessageText = `${errMessageText}: ${macroMatch[1]}`
```

Strauman replied:
Whoop! Didn't know this was possible 🥇 🔢

thomasjo commented on spec/process-manager-spec.js:
```diff
@@ -38,7 +38,7 @@ describe('ProcessManager', () => {
     processManager.executeChildProcess(constructCommand('new.tex'), { allowKill: true }).then(result => { newKilled = true })
 
-    waitsFor(() => oldKilled, 5000)
+    waitsFor(() => oldKilled, 10000)
```
Let's try removing these; I'm hoping whatever the underlying issue on AppVeyor was, that is has been resolved by now. Either way, your changes should not cause any errors in these tests, so if these are still timing out, that should be addressed separately.

Strauman replied:
Done!

thomasjo reviewed (changes requested):
I would really love it if we had a test covering this. The easiest way would probably be to use one of the log files in the fixtures folder (e.g. `file.log`) and augment it to contain an undefined control sequence error. Alternatively create a file which produces that error, and alter the locally generated log file by censoring local file paths, etc, like we've done for existing fixtures.

Having a test covering at least the "happy path" of this new logic ensures correct behavior, guards against regressions, and what not. Having tests that cover cases where a macro is not found would be amazing, but not strictly required.

Aerijo committed:
Update dbus to remove flatmap vulnerability

thomasjo committed:
Prepare 0.50.1 release

Aerijo committed:
Add relative path support (#499)

Fixes #480

thomasjo committed:
Prepare 0.50.2 release

Strauman committed:
Merge branch 'control-sequence-name' of github.com:Strauman/atom-latex into control-sequence-name

Strauman commented:
> Having tests that cover cases where a macro is not found would be amazing, but not strictly required.

I'll see if I get the time.

Strauman committed:
Reset timeout on process-manager-spec.js

Due to continuously timing out the test on this, I added a longer timeout.
This is now removed per request:
https://github.com/thomasjo/atom-latex/pull/496#discussion_r238993895

Strauman committed:
Misc minor review request changes

- const where const is due
- Better variable names
- More canonical inline variables in string

Also a few extra comments

Strauman committed:
Out of bounds test on lines when macro not found

As per review request, now checking that there exists at least one line
below "Unknown control sequence" error before starting to parse it.

Strauman committed:
Fixed linting error
